### PR TITLE
Logo Customization

### PR DESCRIPTION
**Pull request by princemaple**
close https://github.com/firezone/product/issues/412
close https://github.com/firezone/product/issues/411
close https://github.com/firezone/firezone/issues/525

image
---
image
image
image

### CODE REVIEW

File: apps/fz_http/lib/fz_http_web/live/setting_live/customization_live.ex [hunk 1]
```diff
@@ -0,0 +1,78 @@
+defmodule FzHttpWeb.SettingLive.Customization do
+  @moduledoc """
+  Manages the app customizations.
+  """
+  use FzHttpWeb, :live_view
+
+  alias FzHttp.Conf
+
+  @max_logo_size 256 * 1024
```
Comment (jamilbk) [inline apps/fz_http/lib/fz_http_web/live/setting_live/customization_live.ex]: Wondering if 256 KB might be too small. Maybe bump to 1 MB?

File: apps/fz_http/lib/fz_http_web/live/setting_live/customization_live.ex [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+            logo: %{"data" => data, "type" => entry.client_type}
+          })
+
+        {:ok, config}
+      end)
+
+    {:noreply, assign(socket, :config, config)}
+  end
+
+  defp error_to_string(:too_large), do: "The file is too large"
```
Comment (jamilbk) [inline apps/fz_http/lib/fz_http_web/live/setting_live/customization_live.ex]:
```suggestion
  defp error_to_string(:too_large), do: "The file exceeds the maximum size of X."
```

File: .credo.exs [hunk 3]
```diff
@@ -0,0 +1,208 @@
+# This file contains the configuration for Credo and you are probably reading
```
Comment (jamilbk) [inline .credo.exs:1]: Thanks for adding this!

**Review by jamilbk — CHANGES_REQUESTED** [review]
Very minor suggestions.

File: apps/fz_http/lib/fz_http_web/live/setting_live/customization_live.ex (hunk 1 above)
Comment (princemaple) [reply apps/fz_http/lib/fz_http_web/live/setting_live/customization_live.ex]: I almost wanted to limit it to under 100KB.. 😅 I really didn't want to send that much data over db connection.

Do you think a logo should be that large?

Comment (jamilbk) [reply apps/fz_http/lib/fz_http_web/live/setting_live/customization_live.ex]: Naw, I really hope not. I just know that we'll get users complaining that they can't upload their high DPI amazing company logo... lol.

Hmm I was afraid of that... the DB solution is so easy but hard to cache. While uploading to the filesystem is hard (and creates scalability problems) but can be cached by nginx handily.

Comment (princemaple) [reply apps/fz_http/lib/fz_http_web/live/setting_live/customization_live.ex]: The middle ground is of course caching in memory with persistent terms

Comment (jamilbk) [reply apps/fz_http/lib/fz_http_web/live/setting_live/customization_live.ex]: > The middle ground is of course caching in memory with persistent terms

Sounds good. This works for now 👍

**Comment by princemaple** [comment]
Flaky test 😿 will try changing the assertion tomorrow

**Comment by jamilbk** [comment]
@princemaple Almost forgot, would it be possible to add this to the `ping` telemetry? Trying to get better at remembering to add new features there.

```elixir
defmodule FzHttp.Telemetry do
  def ping_data do
    ...
    custom_logo: false, # logo is nil
    # or
    custom_logo: :url, # custom logo is specified with URL
    # or
    custom_logo: :file, # custom logo is specified with a file
  end
end
```

**Comment by jamilbk** [comment]
I don't think admins will be changing these conf keys very often, so `persistent_term` should work well. `persistent_term` would be a really great fit for moving boot-time app config into as well since those never change.

**Pull request merged**